Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -2448,6 +2462,42 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
     });
 }
 
+RefPtr<TextIndicator> UnifiedPDFPlugin::textIndicatorForSelection(OptionSet<WebCore::TextIndicatorOption> options, WebCore::TextIndicatorPresentationTransition transition)
+{
+    if (auto selectionBounds = selectionBoundsForFirstPageInDocumentSpace(m_currentSelection)) {
+        auto rectInDocumentCoordinates = *selectionBounds;
+        auto rectInContentsCoordinates = convertFromDocumentToContents(rectInDocumentCoordinates);
+        auto rectInRootViewCoordinates = convertFromPluginToRootView(convertFromDocumentToPlugin(rectInDocumentCoordinates));
+
+        float deviceScaleFactor = this->deviceScaleFactor();
+
+        auto buffer = ImageBuffer::create(rectInRootViewCoordinates.size(), RenderingPurpose::ShareableSnapshot, deviceScaleFactor, DestinationColorSpace::SRGB(), PixelFormat::BGRA8, { }, nullptr);
+        if (!buffer)
+            return nullptr;
+
+        auto& context = buffer->context();
+        context.translate(-rectInContentsCoordinates.location());
+        context.scale(m_scaleFactor);
+        context.translate(centeringOffset());
+        paintPDFContent(context, rectInContentsCoordinates, PaintingBehavior::PageContentsOnly);
+
+        TextIndicatorData data;
+        data.contentImage = BitmapImage::create(ImageBuffer::sinkIntoNativeImage(WTFMove(buffer)));
+        data.contentImageScaleFactor = deviceScaleFactor;
+        data.contentImageWithoutSelection = data.contentImage;
+        data.contentImageWithoutSelectionRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.selectionRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.textBoundingRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.textRectsInBoundingRectCoordinates = { FloatRect { 0, 0, static_cast<float>(rectInRootViewCoordinates.width()), static_cast<float>(rectInRootViewCoordinates.height()) } };
+        data.presentationTransition = transition;
+        data.options = options;
```

Comment:
Is this a pretty common set of fields that are set together for text indicators? If so maybe we could do some TextIndicator cleanup late to help with this case or we could just create a helper for the `UnifiedPDFPlugin``

Reply:
This is the helper. There are some improvements we can make here (vending one with the selection but still no background on `contentImage` and without the selection on `contentImageWithoutSelection`, etc.) to improve transitions, but it's a start.